python/python/tests/test_pubsub.py

```diff
@@ -0,0 +1,1800 @@
+# Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0
+
+from __future__ import annotations
+
+import asyncio
+from enum import IntEnum
+from typing import Any, Dict, List, Optional, Set, Tuple, Union
+
+import pytest
+from glide.async_commands.core import InfoSection
+from glide.config import (
+    ClusterClientConfiguration,
+    ProtocolVersion,
+    RedisClientConfiguration,
+)
+from glide.constants import OK
+from glide.exceptions import WrongConfiguration
+from glide.redis_client import RedisClient, RedisClusterClient, TRedisClient
+from tests.conftest import create_client
+from tests.utils.utils import check_if_server_version_lt, get_random_string
+
+
+class PubSubChannelModes(IntEnum):
+    Exact = 0
+    """ Use exact channel names """
+    Pattern = 1
+    """ Use channel name patterns """
+    Sharded = 2
+    """ Use sharded pubsub """
+
+
+class MethodTesting(IntEnum):
+    """
+    Enumeration for specifying the method of Pub/Sub subscription.
+    """
+
+    Async = 0
+    Sync = 1
+    Callback = 2
+
+
+async def setup_clients(
+    request,
+    cluster_mode,
+    pub_sub,
+    pub_sub2: Optional[Any] = None,
+    protocol: ProtocolVersion = ProtocolVersion.RESP3,
+) -> Tuple[
+    Union[RedisClient, RedisClusterClient], Union[RedisClient, RedisClusterClient]
+]:
+    "Sets up clients for testing purposes."
+    c_pubsub, s_pubsub = None, None
+    c2_pubsub, s2_pubsub = None, None
+    if cluster_mode:
+        c_pubsub = pub_sub
+        c2_pubsub = pub_sub2
+    else:
+        s_pubsub = pub_sub
+        s2_pubsub = pub_sub2
+
+    publishing_client = await create_client(
+        request,
+        cluster_mode=cluster_mode,
+        cluster_mode_pubsub=c2_pubsub,
+        standalone_mode_pubsub=s2_pubsub,
+        protocol=protocol,
+    )
+    listening_client = await create_client(
+        request,
+        cluster_mode=cluster_mode,
+        cluster_mode_pubsub=c_pubsub,
+        standalone_mode_pubsub=s_pubsub,
+        protocol=protocol,
+    )
+    return publishing_client, listening_client
+
+
+async def get_message_by_method(
+    method: MethodTesting,
+    client: TRedisClient,
+    messages: Optional[
+        List[Union[RedisClient.PubSubMsg, RedisClusterClient.PubSubMsg]]
+    ] = None,
+    index: Optional[int] = None,
+):
+    if method == MethodTesting.Async:
+        return await client.get_pubsub_message()
+    elif method == MethodTesting.Sync:
+        return client.try_get_pubsub_message()
+    assert messages and (index is not None)
+    return messages[index]
```

Comment:
Per commend at line 1515, please ensure you are not assuming the ordering of the messages with the regards to the publish calls - lets discuss it